util/argo/resource_tracking.go

```diff
@@ -167,8 +167,54 @@ func (rt *resourceTracking) SetAppInstance(un *unstructured.Unstructured, key, v
 		}
 		return nil
 	default:
-		return setAppInstanceAnnotation()
+		if err := setAppInstanceAnnotation(); err != nil {
+			return err
+		}
+	}
+
+	// Record how we tracked the resource (label, annotation, or both),
+	// so that we can later remove only those when needed.
+	// This is useful to remove the all known tracking label/annotations
+	// set by Argo CD especially when generating hydrated manifest.
+	return kube.SetAppInstanceAnnotation(un, v1alpha1.AnnotationKeyTrackingMethod, string(trackingMethod))
+}
+
+func (rt *resourceTracking) RemoveAppInstance(un *unstructured.Unstructured) error {
+	trackingMethod, _, _ := unstructured.NestedString(un.Object, "metadata", "annotations", v1alpha1.AnnotationKeyTrackingMethod)
+
+	switch v1alpha1.TrackingMethod(trackingMethod) {
```

Comment:
Each of the cases should return nil early if there's no error, like in `SetAppInstance`.

Reply:
Please see now.